batcher/aligned-batcher/src/lib.rs

```diff
@@ -455,206 +341,177 @@ impl Batcher {
+        if !self.check_min_balance(&addr, proofs_in_batch + 1).await {
+            std::mem::drop(batch_state_lock);
             send_message(
                 ws_conn_sink.clone(),
-                ValidityResponseMessage::InvalidSignature,
+                ValidityResponseMessage::InsufficientBalance(addr),
             )
             .await;
-            Ok(()) // Send error message to the client and return
+            return Ok(());
         }
-    }
 
-    // Checks user has sufficient balance
-    // If user has sufficient balance, increments the user's proof count in the batch
-    async fn check_user_balance_and_increment_proof_count(&self, addr: &Address) -> bool {
-        if self.user_balance_is_unlocked(addr).await {
-            return false;
-        }
-        let mut batch_state = self.batch_state.lock().await;
+        let user_nonce = batch_state_lock.get_user_nonce(&addr).await;
+        let expected_nonce = if let Some(cached_nonce) = user_nonce {
+            cached_nonce
+        } else {
+            let ethereum_user_nonce = match self.get_user_nonce_from_ethereum(addr).await {
+                Ok(ethereum_user_nonce) => ethereum_user_nonce,
+                Err(e) => {
+                    error!(
+                        "Failed to get user nonce from Ethereum for address {addr:?}. Error: {e:?}"
+                    );
+                    std::mem::drop(batch_state_lock);
+                    send_message(ws_conn_sink.clone(), ValidityResponseMessage::InvalidNonce).await;
+                    return Ok(());
+                }
+            };
 
-        let user_proofs_in_batch = batch_state.get_user_proof_count(addr) + 1;
+            batch_state_lock.update_user_nonce(&addr, ethereum_user_nonce);
+            ethereum_user_nonce
+        };
 
-        let user_balance = self.get_user_balance(addr).await;
+        if expected_nonce < msg_nonce {
+            std::mem::drop(batch_state_lock);
+            warn!("Invalid nonce for address {addr}, had nonce {expected_nonce:?} < {msg_nonce:?}");
+            send_message(ws_conn_sink.clone(), ValidityResponseMessage::InvalidNonce).await;
+            return Ok(());
+        }
 
-        let min_balance = U256::from(user_proofs_in_batch) * U256::from(MIN_FEE_PER_PROOF);
-        if user_balance < min_balance {
-            return false;
+        // In this case, the message might be a replacement one. If it is valid,
+        // we replace the old entry with the new from the replacement message.
+        if expected_nonce > msg_nonce {
+            info!("Possible replacement message received: Expected nonce {expected_nonce:?} - message nonce: {msg_nonce:?}");
+            self.handle_replacement_message(
+                batch_state_lock,
+                nonced_verification_data,
+                ws_conn_sink.clone(),
+                client_msg.signature,
+                addr,
+            )
```

Comment:
Let's make sure to also drop the lock inside this function before responding to the network.

Reply:
forgot to answer here, but same as in the other cases, the `handle_replacemente_message` function takes ownership of the lock and drops it